docs/fileformat-prose.rst

```diff
@@ -0,0 +1,180 @@
+
+Dataframes can be identified from other groups by their attributes:
+
+.. code:: python
+
+   >>> dict(f["obs"].attrs)
+   {'_index': 'Cell',
+    'column-order': array(['sample', 'cell_type', 'n_genes_by_counts',
+          'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts',
+          'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes',
+          'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes',
+          'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito',
+          'label_by_score'], dtype=object),
+    'encoding-type': 'dataframe',
+    'encoding-version': '0.1.0'}
+
+These attributes identify the column used as an index, the order of the original columns, and some type information.
+
+.. code:: python
+
+   >>> f["obs"].visititems(lambda k, v: print(f"{k}: {v}"))
+   Cell: <HDF5 dataset "Cell": shape (38410,), type "|O">
+   __categories: <HDF5 group "/obs/__categories" (3 members)>
+   __categories/cell_type: <HDF5 dataset "cell_type": shape (22,), type "|O">
+   __categories/label_by_score: <HDF5 dataset "label_by_score": shape (16,), type "|O">
+   __categories/sample: <HDF5 dataset "sample": shape (41,), type "|O">
+   cell_type: <HDF5 dataset "cell_type": shape (38410,), type "|i1">
+   label_by_score: <HDF5 dataset "label_by_score": shape (38410,), type "|i1">
+   log1p_n_genes_by_counts: <HDF5 dataset "log1p_n_genes_by_counts": shape (38410,), type "<f8">
+   log1p_total_counts: <HDF5 dataset "log1p_total_counts": shape (38410,), type "<f4">
+   log1p_total_counts_mito: <HDF5 dataset "log1p_total_counts_mito": shape (38410,), type "<f4">
+   n_genes_by_counts: <HDF5 dataset "n_genes_by_counts": shape (38410,), type "<i4">
+   pct_counts_in_top_100_genes: <HDF5 dataset "pct_counts_in_top_100_genes": shape (38410,), type "<f8">
+   pct_counts_in_top_200_genes: <HDF5 dataset "pct_counts_in_top_200_genes": shape (38410,), type "<f8">
+   pct_counts_in_top_500_genes: <HDF5 dataset "pct_counts_in_top_500_genes": shape (38410,), type "<f8">
+   pct_counts_in_top_50_genes: <HDF5 dataset "pct_counts_in_top_50_genes": shape (38410,), type "<f8">
+   pct_counts_mito: <HDF5 dataset "pct_counts_mito": shape (38410,), type "<f4">
+   sample: <HDF5 dataset "sample": shape (38410,), type "|i1">
+   total_counts: <HDF5 dataset "total_counts": shape (38410,), type "<f4">
+   total_counts_mito: <HDF5 dataset "total_counts_mito": shape (38410,), type "<f4">
+
+Categorical series can be identified by the presence of the attribute `"categories"`, which contains a pointer to their categorical values:
+
+*Note:* as `zarr` does not have reference objects, in zarr files the `categories` attribute is an absolute path to the category values.
+
+.. code:: python
+
+   >>> dict(f["obs/cell_type"].attrs)
+   {'categories': <HDF5 object reference>}
+
+Other values:
+-------------
+
+Mappings
+~~~~~~~~
+
+Mappings are stored as native groups in an `h5ad` file.
```

Comment:
I just realized: You said we can use any kind of mapping keys for `AlignedArray`s, but using hdf5 group names to represent those keys that’s wrong.

Reply:
Did I say that? I think I've said anything could work in memory, but only strings would work on disk.